aten/src/ATen/native/cudnn/BatchNorm.cpp

```diff
@@ -195,41 +201,103 @@ std::tuple<Tensor, Tensor, Tensor, Tensor> cudnn_batch_norm(
+        input->data_ptr(),
+        idesc.desc(),
+        output->data_ptr(),
+        wdesc.desc(),
+        weight->data_ptr(),
+        bias->data_ptr(),
+        exponential_average_factor,
+        at::maybe_data_ptr(running_mean),
+        at::maybe_data_ptr(running_var),
+        epsilon,
+        save_mean.data_ptr(),
+        save_var.data_ptr()));
 #endif // CUDNN_VERSION >= 7400
   } else {
-    reserve = at::empty({0}, input->options().dtype(kByte));
     AT_CUDNN_CHECK(cudnnBatchNormalizationForwardInference(
-      handle, mode, &one, &zero,
-      idesc.desc(), input->data_ptr(),
-      idesc.desc(), output->data_ptr(),
-      wdesc.desc(),
-      weight->data_ptr(),
-      bias->data_ptr(),
-      running_mean->data_ptr(),
-      running_var->data_ptr(),
-      epsilon));
+        handle,
+        mode,
+        &one,
+        &zero,
+        idesc.desc(),
+        input->data_ptr(),
+        idesc.desc(),
+        output->data_ptr(),
+        wdesc.desc(),
+        weight->data_ptr(),
+        bias->data_ptr(),
+        running_mean->data_ptr(),
+        running_var->data_ptr(),
+        epsilon));
   }
 
   // save_mean and save_var can be undefined
   // If this causes problems, we can initialize them to empty tensors
   // of the correct type
-  return std::tuple<Tensor, Tensor, Tensor, Tensor>{output_t, save_mean, save_var, reserve};
+  return std::tuple<Tensor&, Tensor&, Tensor&, Tensor&>{
+      output_t, save_mean, save_var, reserve};
 }
 
+std::tuple<Tensor, Tensor, Tensor, Tensor> cudnn_batch_norm(
+    const Tensor& input_t,
+    const Tensor& weight_t,
+    const c10::optional<Tensor>& bias_t_opt,
+    const c10::optional<Tensor>& running_mean_t_opt,
+    const c10::optional<Tensor>& running_var_t_opt,
+    bool training,
+    double exponential_average_factor,
+    double epsilon) {
+  TensorArg input{input_t, "input", 1};
+  auto output_t =
+      at::empty_like(*input, input->options(), input->suggest_memory_format());
+
+  auto handle = getCudnnHandle();
+  auto dataType = getCudnnDataType(*input);
+  TensorDescriptor idesc{*input, 4}; // input descriptor
+  Tensor save_mean, save_var, reserve;
+  cudnnBatchNormMode_t mode = getCudnnBatchNormMode(
+      training, input->suggest_memory_format(), input->dim());
+
+  if (training) {
+    int64_t num_features = input_t.size(1);
+    save_mean = at::empty({num_features}, weight_t.options());
+    save_var = at::empty({num_features}, weight_t.options());
+
+#if CUDNN_VERSION >= 7400
+    auto op = CUDNN_BATCHNORM_OPS_BN;
+    size_t workspace_size;
+    // get the reserved size and allocate as tensor
+    size_t reserve_size;
+    AT_CUDNN_CHECK(cudnnGetBatchNormalizationTrainingExReserveSpaceSize(
+        handle, mode, op, nullptr, idesc.desc(), &reserve_size));
+    reserve = at::empty(reserve_size, input->options().dtype(kByte));
+#endif // CUDNN_VERSION >= 7400
+  }
+
+  std::tie(output_t, save_mean, save_var, reserve) = cudnn_batch_norm_out(
```

Comment:
Would linter complain if we skip the assignment to std::tie?

Reply:
C++ doesn't do this kind of unpacking before [`std=c++17`](https://en.cppreference.com/w/cpp/language/structured_binding) and our cmake sets [`std=c++14`](https://github.com/pytorch/pytorch/blob/master/CMakeLists.txt#L40)

Reply:
oh duh whoops. will change. thanks!